spacy/tokens/doc.pyx

```diff
@@ -1381,7 +1366,7 @@ cdef int _set_lr_kids_and_edges(TokenC* tokens, int length, int loop_count) exce
     # Get sentence start positions according to current state
     sent_starts = set()
     for i in range(length):
-        if tokens[i].head == 0 and tokens[i].dep != 0:
+        if tokens[i].head == 0:
```

Comment:
Is this right?

Reply:
This is actually a duplicate change from #6044. I had tried to replace the code in `Token.head` with `set_children_from_heads` before, but always ran into test failures and weird behavior. It had to do with gradually constructing or modifying a doc in a loop, which was done in some of the tests, like this:

```
for i in range(length):
  doc[i].dep_ = deps[i]
  doc[i].head = doc[heads[i]]
```

Depending on whether the dep or head is set first, you got different behavior in the lr_edges/kids/sent_start because of this. I really think `set_children_from_heads` should not care what the deps are. It should only care about the shape of the unlabelled graph. The additional code in `Token.head` also didn't adjust `is_sent_start` correctly, just the lr_edges/kids. I think maintaining only `set_children_from_heads` is going to be an overall improvement, even though it will be slightly slower because it goes through the whole doc.

Reply:
The problem is you can only really set the heads per token, so it's pretty unfortunate if looping over the document and changing the heads is `O(n**2)`.

I'll reluctantly agree to merge this because I agree that the code duplication is really bad. But is there a way we can break out the loop body of `set_children_from_heads` so that we can call it with only one token?

I'd rather you didn't spend a lot of time on this just now, so if it's tricky just leave a todo and we can come back to it.

Reply:
I think it would be pretty easy to limit the range of the tokens adjusted by `Token.head`. I think you could go from the leftmost left edge of either root to the rightmost right edge of either root? Does that sound right?

Reply:
Yes that does make sense. Do you want to try that? I think it sounds like a good approach.